frame/data-feed/src/lib.rs

```diff
@@ -0,0 +1,721 @@
+		pub fn add_provider(origin, new_one: <T::Lookup as StaticLookup>::Source) -> DispatchResult {
+			T::DispatchOrigin::try_origin(origin).map(|_| ()).or_else(ensure_root)?;
+			let new_one = T::Lookup::lookup(new_one)?;
+
+			ActiveProviders::<T>::try_mutate(|v| -> DispatchResult {
+				if !v.contains(&new_one) {
+					if v.len() + 1 > MAX_KEY_LEN {
+						Err(Error::<T>::ExceedLimit)?
+					}
+
+					v.push(new_one.clone());
+				}
+				Ok(())
+			})?;
+			Self::deposit_event(RawEvent::AddProvider(new_one));
+			Ok(())
+		}
+
+		/// Remove a provider from current provider collection.
+		#[weight = T::WeightInfo::remove_provider()]
+		pub fn remove_provider(origin, who: <T::Lookup as StaticLookup>::Source) -> DispatchResult {
+			T::DispatchOrigin::try_origin(origin).map(|_| ()).or_else(ensure_root)?;
+			let who = T::Lookup::lookup(who)?;
+
+			ActiveProviders::<T>::mutate(|v| {
+				v.retain(|accountid| accountid != &who);
+			});
+			Self::deposit_event(RawEvent::RemoveProvider(who));
+			Ok(())
+		}
+
+		fn on_finalize(_n: T::BlockNumber) {
+			for (key, info) in Infos::<T>::iter() {
+				if  _n % info.schedule == Zero::zero() {
+					Self::calc(key, info);
+				}
+			}
+		}
+
+		fn offchain_worker(block_number: T::BlockNumber) {
+			debug::native::info!("Initialize offchain workers!");
+
+			for key in Self::all_keys() {
+				let res = Self::fetch_and_send_signed(&key);
+				if let Err(e) = res {
+					debug::error!("Error: {}", e);
+				}
+			}
+		}
+	}
+}
+
+impl<T: Trait> Module<T> {
+	fn feed_data_impl(who: T::AccountId, key: StorageKey, value: DataType) -> DispatchResult {
+		if !Self::all_keys().contains(&key) {
+			Err(Error::<T>::InvalidKey)?
+		}
+		let info: Info<_> = Self::infos(&key).ok_or(Error::<T>::InvalidKey)?;
+		if value.number_type() != info.number_type {
+			Err(Error::<T>::InvalidValue)?
+		}
+		debug::info!("Feeding data to {:?}", key);
+		DataFeeds::<T>::try_mutate_exists(key.clone(), who.clone(), |data| {
+			match data {
+				Some(data) => {
+					let mut new_data = [Default::default(); RING_BUF_LEN];
+					new_data[0] = value;
+					// move data buf to new_data buf, and drop last item
+					new_data[1..].copy_from_slice(&data[0..(data.len() - 2)]);
+					*data = new_data;
+				}
+				None => {
+					*data = Some([value; RING_BUF_LEN]);
+				}
+			}
+			Self::deposit_event(RawEvent::FeedData(who, key, value));
+			Ok(())
+		})
+	}
+
+	fn calc_impl<N: Saturating + CheckedDiv + Copy, F: FnOnce(usize) -> N>(
+		numbers: &[N],
+		zero: N,
+		convert: F,
+		op: Operations,
+	) -> N {
+		match op {
+			Operations::Sum => {
+				let sum = numbers.iter().fold(zero, |a, b| a.saturating_add(*b));
+				sum
+			}
+			Operations::Average => {
+				if numbers.is_empty() {
+					zero
+				} else {
+					let sum = numbers.iter().fold(zero, |a, b| a.saturating_add(*b));
+					// let n = N::from(numbers.len() as u128);
+					let n = convert(numbers.len());
+					sum.checked_div(&n).unwrap_or(zero)
```

Comment:
This is flawed.

If you hit some saturating point, you need to count then how many values you have and only divide by that amount.

Or we need to in general be more clever here, but this current logic is not safe.

Reply:
I think in general with this calc impl, there should be some result returned rather than the value.

It should be okay to return `None` value, which then triggers some other processes to resolve what the actual value should be.

Reply:
fixed by you suggesion